### ISSUE TITLE
org.hibernate.bytecode.enhance.spi.EnhancementException: Enhancement of [org.acme.domain.MyEntity] failed

### ISSUE THREAD

**Issue by KaiSuchomel** [quarkusio/quarkus#44863]
### Describe the bug

Exception occurs during Build:
` org.hibernate.bytecode.enhance.spi.EnhancementException: Enhancement of [org.acme.domain.MyEntity] failed because no field named [modelType] could be found for property accessor method [getModelType]. To fix this, make sure all property accessor methods have a matching field.`

### Expected behavior

The Build should succeed.

### Actual behavior

When building the project:
ERROR] Caused by: java.util.concurrent.ExecutionException: org.hibernate.bytecode.enhance.spi.EnhancementException: Enhancement of [org.acme.domain.MyEntity] failed because no field named [modelType] could be found for property accessor method [getModelType]. To fix this, make sure all property accessor methods have a matching field.

### How to Reproduce?

Reproducer: https://github.com/KaiSuchomel/quarkus-test/tree/hibernate
Just build the project with maven.
Simple Entity:
```Java
@Entity
public class MyEntity {

    @Id
    private UUID id;

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public void getXXX() {
    }

    @PrePersist
    public void setId(){
        id = UUID.randomUUID();
    }
    
    public String getModelType() {
        return "MyEntity";
    }
}
```
After changing the PrePersist setter to a seperate name the build is working:
```Java
@Entity
public class MyEntity {

    @Id
    private UUID id;

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public void getXXX() {
    }

    @PrePersist
    public void prePersistSetId(){
        id = UUID.randomUUID();
    }
    
    public String getModelType() {
        return "MyEntity";
    }
}
```

### Output of `uname -a` or `ver`

Darwin Mac-C6VD4H4XT3 24.1.0 Darwin Kernel Version 24.1.0: Thu Oct 10 21:00:32 PDT 2024; root:xnu-11215.41.3~2/RELEASE_ARM64_T6030 arm64

### Output of `java -version`

openjdk version "21.0.5" 2024-10-15 OpenJDK Runtime Environment Homebrew (build 21.0.5) OpenJDK 64-Bit Server VM Homebrew (build 21.0.5, mixed mode, sharing)

### Quarkus version or git rev

3.17.2 --> hibernate-core 6.6.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven home: /opt/homebrew/Cellar/maven/3.9.9/libexec Java version: 21.0.5, vendor: Homebrew, runtime: /opt/homebrew/Cellar/openjdk@21/21.0.5/libexec/openjdk.jdk/Contents/Home Default locale: de_DE, platform encoding: UTF-8 OS name: "mac os x", version: "15.1.1", arch: "aarch64", family: "mac"

### Additional information

@yrodiere Can you please check. Thx.
See: https://github.com/hibernate/hibernate-orm/commit/b5221e2ec65d0118fb04fcbc496e599066014b7a#diff-5f775bce3db8bc5527104f2eba37fef522276d3bed00a3276545528ab233c380R16

**Comment by mbladel** [comment]
Hello @KaiSuchomel, thank you for reporting the issue and providing a simple mapping which reproduces your problem.

The error you're facing is caused by a recent change in Hibernate's bytecode-enhancement logic: https://hibernate.atlassian.net/browse/HHH-16572. The new check is needed because support for bytecode enhanced entities with `AccessType.PROPERTY` (using getter/setter methods instead of fields) does not function properly unless the field names exactly match the method names - in Quarkus this can be especially problematic since enhancement is done at build-time and we expect all entities to be enhanced at runtime.

Using `setX`, where `X` is a persistent property name, as the method name in your entity classes is generally not a good idea, especially when using bytecode-enhancement, as the method could be mistaken as a setter for that field. So the "workaround" you found is more than valid and I would suggest keeping that as-is.

Regardless, I believe the check should in this case insce:

1. The entity is using `AccessType.FIELD`, since the `@Id` annotation is placed on the `id` field;
2. The pre-persist method `setId` corresponds to an actually existing field

We should open an issue on Hibernate ORM's [Jira](https://hibernate.atlassian.net/browse/HHH) to keep track of this and possibly improve the check.

**Comment by geoand** [comment]
@mbladel is https://github.com/quarkusio/quarkus/issues/44863 caused by the same thing?

**Comment by mbladel** [comment]
@geoand I believe so, yes, I've added a link to that issue and its reproducer to the ORM Jira - thanks for the ping.

**Comment by wcicola** [comment]
I've recently encountered a similar issue that seems to be related, but haven't yet seen a report of this exact manifestation of it either here or on the Hibernate Jira.

If an entity:

1. Is mapped using `AccessType.PROPERTY`,
2. Extends a class with fields that are accessible to subclasses, and
3. Tries to map those accessible superclass fields using annotations on its own getters/setters

Then the build fails also with a similar EnhancementException.

I can put together a full repro project and/or post on the Hibernate Jira if it would help, but the only required code for reproduction is a Quarkus 3.17.2+ project using `quarkus-hibernate-orm` and with these two classes:

```
package test.entity;

import jakarta.persistence.MappedSuperclass;

@MappedSuperclass
public abstract class AbstractEntity {

	protected Long oid;
}
```


```
package test.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;

@Entity
public class EntityTest extends AbstractEntity {

	@Id
	@Column(name = "ID")
	public Long getOid() {

		return oid;
	}

	public void setOid(Long oid) {

		this.oid = oid;
	}
}
```

The result of `mvn package`:
```
[ERROR] Caused by: org.hibernate.bytecode.enhance.spi.EnhancementException: Enhancement of [test.entity.EntityTest] failed because no field named [oid] could be found for property accessor method [getOid]. To fix this, make sure all property accessor methods have a matching field.
```

Basically, the bytecode enhancement process seems to now be only considering fields declared in the same class that contains the getters/setters, whereas it previously detected inherited fields without issue. The same code builds fine using Quarkus 3.17.0 or lower.

**Comment by yrodiere** [comment]
@wcicola the present issue was fixed in https://hibernate.atlassian.net/browse/HHH-18904, which means it's fixed in Quarkus 3.17.6 (through #45235). I unfortunately forgot to close issues fixed in the latest ORM version, I'll get to it right now.

Could you please test this with Quarkus 3.17.6, and if your problem still occurs, open a new issue? Preferably, directly on hibernate.atlassian.net and with a reproducer based on [this](https://github.com/hibernate/hibernate-test-case-templates/blob/main/orm/hibernate-orm-6/src/test/java/org/hibernate/bugs/QuarkusLikeORMUnitTestCase.java).

Thank you.

**Comment by wcicola** [comment]
@yrodiere  @mbladel  Many thanks! I'll get that put together and posted today.

**Comment by dhufnagel** [comment]
We are discovering this issue in 3.21.2 again. It worked until 3.21.1 and started to fail in 3.21.2. Anyone else rediscovering this?

**Comment by KaiSuchomel** [comment]
> We are discovering this issue in 3.21.2 again. It worked until 3.21.1 and started to fail in 3.21.2. Anyone else rediscovering this?

In our Environment the deployment is failing with the error again with quarkus 3.21.2. Running with quarkus 3.32.1.
@dhufnagel Have you a cretaed a reproducer?
May be the https://github.com/quarkusio/quarkus/pull/47089 is causing the issue.

**Comment by KaiSuchomel** [comment]
@dhufnagel @mbladel 
I have enhance the reproducer: https://github.com/KaiSuchomel/quarkus-test/tree/hibernate
With quarkus 3.21.1 everything is running, quarkus 3.21.2 is failing with:
`[ERROR] Caused by: org.hibernate.bytecode.enhance.spi.EnhancementException: Enhancement of [org.acme.domain.RevisionBase] failed because no field named [id] could be found for property accessor method [getId]. To fix this, make sure all property accessor methods have a matching field.`

**Comment by dhufnagel** [comment]
> > We are discovering this issue in 3.21.2 again. It worked until 3.21.1 and started to fail in 3.21.2. Anyone else rediscovering this?
> 
> In our Environment the deployment is failing with the error again with quarkus 3.21.2. Running with quarkus 3.32.1.
> @dhufnagel Have you a cretaed a reproducer?
> May be the https://github.com/quarkusio/quarkus/pull/47089 is causing the issue.
> 

I was out of Office and did. Not have the time to create a reproducer. But it seems to be related to the issue @mbladel is mentioning. We have an entity with a getSomething() method, where "something" is not a property but extract information from other properties. So I think in our case we track the mentioned issue.
